### PR TITLE
Refactor/replace generic attribute succession errors with specific ones

### PR DESCRIPTION
**Pull request by Magnus-Kuhn**

### CODE REVIEW

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 1]
```diff
@@ -572,11 +572,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isOwnSharedIdentityAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid own shared identity attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorNotValidOwnSharedIdentityAttribute());
```
Comment (slavistan) [inline packages/consumption/src/modules/attributes/AttributesController.ts:575]: I feel like _is not valid_ is ambiguous and suggests multiple interpretations:
- the attribute is _invalid_, i.e. the predecessor consists of an altogether erroneous payload
- the attribute is not an own shared identity attribute (which is the actual reason for the error here)
- the attribute's `validFrom` or `validTo` values make it in a way "invalid"

I politely suggest you just remove the "Valid" from the error (and add an "is" for legibility) and call it `predecessorIsNotOwnSharedIdentityAttribute`. Same points apply to some of the other errors.

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 2]
```diff
@@ -23,6 +23,25 @@ class Attributes {
         return new CoreError("error.consumption.attributes.successionPeerIsNotOwner", "The peer of the succeeded attribute is not its owner. This may be an attempt of spoofing.");
     }
 
+    public predecessorSourceAttributeInvalid() {
```
Comment (slavistan) [inline packages/consumption/src/consumption/CoreErrors.ts:26]: How about `predecessorSourceAttributeIsNotRepositoryAttribute`? The name is unnecessarily vague. Longer names are alright, if that's what's bothering you :-)

Comment (slavistan) [reply packages/consumption/src/consumption/CoreErrors.ts:26]: Same for `successorSourceAttributeInvalid`.

**Comment by jkoenig134** [comment]
mission version bumps for consumption + runtime

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 3]
```diff
@@ -536,11 +536,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isRepositoryAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid repository attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorIsNotRepositoryAttribute());
         }
 
         if (!successor.isRepositoryAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidSuccessor("Successor is not a valid repository attribute."));
+            return ValidationResult.error(CoreErrors.attributes.successorNotValidRepositoryAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:543]: Analogously to above, I recommend changing it to `successorIsNotRepositoryAttribute`.

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 4]
```diff
@@ -572,11 +572,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isOwnSharedIdentityAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid own shared identity attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorIsNotOwnSharedIdentityAttribute());
         }
 
         if (!successor.isOwnSharedIdentityAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidSuccessor("Successor is not a valid peer shared identity attribute."));
+            return ValidationResult.error(CoreErrors.attributes.successorNotValidOwnSharedIdentityAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:579]: `successorIsNotOwnSharedIdentityAttribute`

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 5]
```diff
@@ -590,28 +590,24 @@ export class AttributesController extends ConsumptionBaseController {
...
+            return ValidationResult.error(CoreErrors.attributes.successorSourceDoesNotSucceedPredecessorSource());
         }
 
         const repositoryAttributeContentMatchesItsSharedAttribute =
             _.isEqual(predecessorSource.content, predecessor.content) && _.isEqual(successorSource.content, successor.content);
         if (!repositoryAttributeContentMatchesItsSharedAttribute) {
-            return ValidationResult.error(
-                CoreErrors.attributes.invalidSuccessionOfOwnSharedIdentityAttribute("Successor source attribute contents don't match successor shared attribute copy.")
-            );
+            return ValidationResult.error(CoreErrors.attributes.sourceContentNotEqualToCopyContent());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:610]: I'd add an "is" here, to be consistent with the other errors.

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 6]
```diff
@@ -643,11 +639,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isOwnSharedRelationshipAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid own shared relationship attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorIsNotOwnSharedIdentityAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:642]: RelationshipAttribute

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 7]
```diff
@@ -643,11 +639,11 @@ export class AttributesController extends ConsumptionBaseController {
...
         }
 
         if (!successor.isOwnSharedRelationshipAttribute(this.identity.address)) {
-            return ValidationResult.error(CoreErrors.attributes.invalidSuccessor("Successor is not a valid own shared relationship attribute."));
+            return ValidationResult.error(CoreErrors.attributes.successorNotValidOwnSharedIdentityAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:646]: `successorIsNotOwnSharedRelationshipAttribute`

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 8]
```diff
@@ -687,11 +683,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isPeerSharedIdentityAttribute()) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid peer shared identity attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorIsNotPeerSharedIdentityAttribute());
         }
 
         if (!successor.isPeerSharedIdentityAttribute()) {
-            return ValidationResult.error(CoreErrors.attributes.invalidSuccessor("Successor is not a valid peer shared identity attribute."));
+            return ValidationResult.error(CoreErrors.attributes.successorNotValidPeerSharedIdentityAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:690]: `successorIsNotPeerSharedIdentityAttribute`

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 9]
```diff
@@ -727,11 +723,11 @@ export class AttributesController extends ConsumptionBaseController {
         });
 
         if (!predecessor.isPeerSharedRelationshipAttribute()) {
-            return ValidationResult.error(CoreErrors.attributes.invalidPredecessor("Predecessor is not a valid peer shared relationship attribute."));
+            return ValidationResult.error(CoreErrors.attributes.predecessorIsNotPeerSharedRelationshipAttribute());
         }
 
         if (!successor.isPeerSharedRelationshipAttribute()) {
-            return ValidationResult.error(CoreErrors.attributes.invalidSuccessor("Successor is not a valid peer shared relationship attribute."));
+            return ValidationResult.error(CoreErrors.attributes.successorNotValidPeerSharedRelationshipAttribute());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:730]: `successorIsNotPeerSharedRelationshipAttribute`

File: packages/consumption/src/modules/attributes/AttributesController.ts [hunk 10]
```diff
@@ -771,9 +767,7 @@ export class AttributesController extends ConsumptionBaseController {
         }
 
         if (typeof successor.succeeds !== "undefined" && !predecessorId.equals(successor.succeeds.toString())) {
-            return ValidationResult.error(
-                CoreErrors.attributes.invalidPredecessor("The predecessor's id and the explicitly set value for the successor's succeeds field don't match.")
-            );
+            return ValidationResult.error(CoreErrors.attributes.differentPredecessorNamedInSuccessor());
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/modules/attributes/AttributesController.ts:770]: I find the term "name" ambiguous here. Maybe better `setPredecessor(Id)DoesNotMatchActualPredecessor(Id)` or similar.

File: packages/consumption/src/modules/attributes/AttributesController.ts (hunk 3 above)
Comment (Magnus-Kuhn) [reply packages/consumption/src/modules/attributes/AttributesController.ts:543]: Is there not some validation happening regarding the successor? I'm with you regarding inserting `Is` though

Comment (Milena-Czierlinski) [reply packages/consumption/src/modules/attributes/AttributesController.ts:543]: This error is returned, if the successor is not a RepositoryAttribute. Hence, calling the error `successorIsNotRepositoryAttribute` seems accurate to me.

Comment (Magnus-Kuhn) [reply packages/consumption/src/modules/attributes/AttributesController.ts:543]: I looked again and have adopted your change, but also renamed `genericValidationError` into `successorIsNotAValidAttribute`

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 11]
```diff
@@ -49,16 +65,51 @@ class Attributes {
...
+    public successorIsNotPeerSharedIdentityAttribute() {
+        return new CoreError("error.consumption.attributes.successorIsNotPeerSharedIdentityAttribute", "Successor is not a peer shared identity attribute.");
+    }
+
+    public successorIsNotOwnSharedRelationshipAttribute() {
+        return new CoreError("error.consumption.attributes.successorIsNotOwnSharedRelationshipAttribute", "Successor is not an own shared relationship attribute.");
+    }
+
+    public successorIsNotPeerSharedRelationshipAttribute() {
+        return new CoreError("error.consumption.attributes.successorIsNotOwnSharedRelationshipAttribute", "Successor is not a peer shared relationship attribute.");
```
Comment (Milena-Czierlinski) [inline packages/consumption/src/consumption/CoreErrors.ts:105]: `peer` in error code

File: package-lock.json
[the remark this answers is not in the document]
Comment (Magnus-Kuhn) [reply package-lock.json:1]: I see - peggy stole my version bumps, I have to bump again

**Review by Milena-Czierlinski — APPROVED** [review]

File: packages/runtime/src/useCases/consumption/attributes/SucceedRelationshipAttributeAndNotifyPeer.ts [hunk 12]
```diff
@@ -41,10 +41,10 @@ export class SucceedRelationshipAttributeAndNotifyPeerUseCase extends UseCase<Su
     protected async executeInternal(request: SucceedRelationshipAttributeAndNotifyPeerRequest): Promise<Result<SucceedRelationshipAttributeAndNotifyPeerResponse>> {
         const predecessor = await this.attributeController.getLocalAttribute(CoreId.from(request.predecessorId));
         if (typeof predecessor === "undefined") {
-            return Result.fail(CoreErrors.attributes.predecessorDoesNotExist());
+            return Result.fail(CoreErrors.attributes.setPredecessorIdDoesNotMatchActualPredecessorId());
```
Comment (slavistan) [inline packages/runtime/src/useCases/consumption/attributes/SucceedRelationshipAttributeAndNotifyPeer.ts:44]: There error here should rather be `predecessorDoesNotExist()`. Or am I missing something?

Comment (Magnus-Kuhn) [reply packages/runtime/src/useCases/consumption/attributes/SucceedRelationshipAttributeAndNotifyPeer.ts]: No, I was just sloppy - fixed it

**Review by Milena-Czierlinski — APPROVED** [review]

File: packages/consumption/src/consumption/CoreErrors.ts [hunk 13]
```diff
@@ -2,10 +2,10 @@ import { ApplicationError } from "@js-soft/ts-utils";
...
-    public genericValidationError(error: any) {
+    public successorIsNotAValidAttribute(error: any) {
         return new CoreError(
-            "error.consumption.attributes.genericValidationError",
-            "Validation failed during creation of object.",
+            "error.consumption.attributes.successorIsNotAValidAttribute",
+            "Validation failed during creation of successor.",
             error,
             undefined,
             error instanceof Error ? error : undefined
```
Comment (slavistan) [inline packages/consumption/src/consumption/CoreErrors.ts:11]: This error differs from the others in that it takes a parameter. E.g. we don't further explain _why_ an attribute is not a valid repository attribute or where a source attribute's content differs from that of its attribute copy. Any reason to break this pattern here?

This is more of a minor nitpick than anything else. I'll let you decide on whether you want to keep the detail or not. I just found it surprising reading through the errors in that your code suggests that the `successorIsNotAValidAttribute` error is in some way "special".

**Review by slavistan — APPROVED** [review]

File: packages/consumption/src/consumption/CoreErrors.ts (hunk 13 above)
Comment (Magnus-Kuhn) [reply packages/consumption/src/consumption/CoreErrors.ts:11]: This error if for the serval errors - I don't know if there is a good way to subdivide those

Comment (slavistan) [reply packages/consumption/src/consumption/CoreErrors.ts:11]: No need to subdivide. Removing the parameter altogether is what I meant.

```ts
    public successorIsNotAValidAttribute() {
        return new CoreError(
            "error.consumption.attributes.successorIsNotAValidAttribute",
            "Validation failed during creation of successor."
        );
    }
```

Comment (Magnus-Kuhn) [reply packages/consumption/src/consumption/CoreErrors.ts:11]: I want to keep the information on what exactly failed - I assume there is some valuable info :)

**Pull request merged**